sql/core/src/main/scala/org/apache/spark/sql/execution/command/AlterTableCommandParser.scala

```diff
@@ -0,0 +1,410 @@
+        // ALTER TABLE table_name NOT STORED AS DIRECTORIES
+        AlterTableSkewed(
+          tableIdent,
+          Nil,
+          Nil,
+          storedAsDirs = false,
+          notSkewed = false,
+          notStoredAsDirs = true)(node.source)
+
+      case (tableSkewed @ Token("TOK_ALTERTABLE_SKEWED", _)) :: _ =>
+        val skewedArgs = tableSkewed match {
+          case Token("TOK_ALTERTABLE_SKEWED", args :: Nil) =>
+            args match {
+              case Token("TOK_TABLESKEWED", skewedCols :: skewedValues :: stored) =>
+                val cols = skewedCols.children.map(n => cleanAndUnquoteString(n.text))
+                val values = skewedValues match {
+                  case Token("TOK_TABCOLVALUE", colVal) =>
+                    Seq(colVal.map(n => cleanAndUnquoteString(n.text)))
+                  case Token("TOK_TABCOLVALUE_PAIR", pairs) =>
+                    pairs.map {
+                      case Token("TOK_TABCOLVALUES", colVals :: Nil) =>
+                        colVals match {
+                          case Token("TOK_TABCOLVALUE", vals) =>
+                            vals.map(n => cleanAndUnquoteString(n.text))
+                        }
+                    }
+                }
+
+                val storedAsDirs = stored match {
+                  case Token("TOK_STOREDASDIRS", Nil) :: Nil => true
+                  case _ => false
+                }
+
+                (cols, values, storedAsDirs)
+            }
+        }
+        val (cols, values, storedAsDirs) = skewedArgs
+        AlterTableSkewed(
+          tableIdent,
+          cols,
+          values,
+          storedAsDirs,
+          notSkewed = false,
+          notStoredAsDirs = false)(node.source)
+
+      case Token("TOK_ALTERTABLE_SKEWED_LOCATION",
+      Token("TOK_SKEWED_LOCATIONS",
+      Token("TOK_SKEWED_LOCATION_LIST", locationMaps) :: Nil) :: Nil) :: _ =>
+        val skewedMaps = locationMaps.map {
+          case Token("TOK_SKEWED_LOCATION_MAP", key :: value :: Nil) =>
+            val k = key match {
+              case Token(const, Nil) => Seq(cleanAndUnquoteString(const))
+              case Token("TOK_TABCOLVALUES", values :: Nil) =>
+                values match {
+                  case Token("TOK_TABCOLVALUE", vals) =>
+                    vals.map(n => cleanAndUnquoteString(n.text))
+                }
+            }
+            (k, cleanAndUnquoteString(value.text))
+        }.toMap
+        AlterTableSkewedLocation(tableIdent, skewedMaps)(node.source)
+
+      case Token("TOK_ALTERTABLE_ADDPARTS", addPartsArgs) :: _ =>
+        val (allowExisting, parts) = addPartsArgs match {
+          case Token("TOK_IFNOTEXISTS", Nil) :: others => (true, others)
+          case _ => (false, addPartsArgs)
+        }
+        val partitions: ArrayBuffer[(Map[String, Option[String]], Option[String])] =
+          new ArrayBuffer()
+        var currentPart: Map[String, Option[String]] = null
+        parts.map {
```

Comment:
We could change the parser rule in such a way that it will nest the locations inside a partition specification.